edx-solutions/edx-platform
# Fixes JS and CSS errors in inline and course-wide Discussion XBlocks

This change fixes a number of Javascript errors and warnings and ensures that the inline DiscussionXBlock, the DiscussionCourseXBlock, and the main Discussion Forum function as on edx:master.

**JIRA tickets**: Implements [YONK-401](https://openedx.atlassian.net/browse/YONK-401).

**Screenshots**:

Inline DiscussionXBlock:
![inline discussionxblock](https://cloud.githubusercontent.com/assets/7556571/18580366/5806c81c-7c3a-11e6-878f-2b99c648234a.png)

DiscussionCourseXBlock (staff view):
![discussion course xblock - staff](https://cloud.githubusercontent.com/assets/7556571/18580377/62906536-7c3a-11e6-9ab4-f084c8cb7a1d.png)

DiscussionCourseXBlock (student view):
![discussion course xblock - student](https://cloud.githubusercontent.com/assets/7556571/18580388/6f123c62-7c3a-11e6-90d6-008586a6bd9b.png)

Discussion Forum:
![discussion forum - student](https://cloud.githubusercontent.com/assets/7556571/18580393/7824d256-7c3a-11e6-9e4f-f533bdfdfcca.png)

**Testing instructions**:

1. Get edx-solutions running on the parent branch, `ziafazal/fix-broken-features`, then checkout this PR's branch `jill/discussion-xblock-bugfix`.
1. Restart Studio and LMS, allowing the static assets to rebuild (`paver devstack lms`, `paver devstack studio`, without `--fast`).
1. As `staff` user, create a course in Studio.
1. Under `Settings > Advanced Settings > Advanced Module List`, add `["discussion-course"]`.
1. Add an inline discussion component via the main "Discussion" component button.
1. Add a course-wide discussion component via "Advanced > Discussion Course".
1. Publish the course.
1. Open up a developer console to watch for Javascript warnings and errors during the following steps.
1. As `staff` user, load the Course Discussion unit.  Ensure that no discussion-related javascript errors or warnings are shown, namely no messages like:

    ```
Unable to load Sequence: DiscussionApp is not defined
backbone.js:1659: Uncaught Error: Backbone.history has already been started
jquery.js:5434: GET http://localhost:8000/c4x/OpenCraft/IDX101/asset/sass_discussion-forum.css fails
```
1. Visit the Inline Discussion unit, and ensure that clicking `Expand discussion` shows the current posts and an Add Post button, instead of an error modal dialog box.
1. As `staff` user, ensure that you can perform the following actions on your posts from within the Inline, Course Discussion, and main Discussion Forum tab:
    * Create Post
    * Reply to Post
    * Add Comment to a Reply
    * Edit a Post, Reply, and Comment.
    * Test [MathJax](http://docs.mathjax.org/en/latest/start.html) functionality in Post, Reply, Comment (and Preview), with e.g.

       ```
There are two solutions to \(ax^2 + bx + c = 0\) and they are
$$x = {-b \pm \sqrt{b^2-4ac} \over 2a}.$$

When $a \ne 0$
```

    * View Discussion User Profile pages by clicking on the user's name.  Posts should render, including MathJax markup.
    * Delete a Post, Reply, and Comment.
    * Open/close a thread/post - in the Instructor panel, add the `staff` user to the `Discussion Admin` or `Discussion Moderator` group to get access to these functions.
    * Follow/unfollow a thread/post
    * Pin/unpin a thread/post
    * Search for posts by clicking the search button, or by hitting Enter in the search box.
    * With the LMS page loaded, stop the LMS service, and try to Create a Post.  Check that the error modal shows up.
1. Login as a student user and register for your course.  You should be able to do the above actions, plus:
    * Report/unreport a post created by the staff user.
    * Vote/unvote a thread/post created by the staff user.

**Author notes and concerns**:

1. Out of scope for this task is to move the code from [`lms.django_comments_client.forum`] to [`lms.discussion`](https://github.com/edx/edx-platform/tree/master/lms/djangoapps/discussion).  cf `edx:master` changes starting from [89f93df](https://github.com/edx/edx-platform/commit/89f93df46e6227058a4c7ef5c00cdb4a456e4583#diff-f0afdf1bdfbfc3a47f6722941269b4a4).
1. Fixing test failures is an ongoing process that is part of the overall rebase project.

**Reviewers**
